### PR TITLE
Ignore deleted events for inexistent packages

### LINKED ISSUE #4463: NPM sync is broken
https://gratipay.slack.com/archives/C36LJJF9V/p1494580201702422

### PR DESCRIPTION
**Pull request by rohitpaulk**
Fixes #4463

### CODE REVIEW

File: tests/py/test_sync_npm.py [hunk 1]
```diff
@@ -85,21 +83,6 @@ def test_even_deletes_package_with_linked_team(self):
         sync_npm.consume_change_stream(self.change_stream(docs), self.db)
         assert self.db.one('select * from packages') is None
 
-
-    def test_picks_up_with_last_seq(self):
```
Comment (whit537) [inline tests/py/test_sync_npm.py]: Is this covered by other tests in a way I'm not seeing?

**Comment by clone1018** [comment]
Do we understand why a package was marked as deleted that we didn't have in our index?

**Comment by whit537** [comment]
From [sentry](https://sentry.io/gratipay/gratipay-com/issues/271411814/) it looks like `adam-test` is the package in question. Could recrawl the changes stream from local to investigate.

File: tests/py/test_sync_npm.py (hunk 1 above)
Comment (rohitpaulk) [inline tests/py/test_sync_npm.py]: This test currently only tested our test function (`change_stream(self, changes)`). The job of picking up from `last_seq` now belongs to `production_change_stream`, which internally calls the CouchDB library, and hence we don't need to test it.

**Comment by rohitpaulk** [comment]
> Could recrawl the changes stream from local to investigate.

Yes, we could do that. However, we haven't seen an error like this across 4 million records - is it really worth the effort? :)

Middle ground would be to merge this to get npm sync back up and running, and maybe log this error out and count logs a week later to decide whether we should investigate.

**Comment by whit537** [comment]
We could also manually insert an `adam-test` package to then be deleted. We will need to learn whether this is something we're going to hit regularly or a one-off fluke.

The risk is that we'll be masking a bug by blanket-ignoring these discrepancies.

**Comment by rohitpaulk** [comment]
> We could also manually insert an adam-test package to then be deleted.

Makes sense, I'm happy to go ahead with this

**Comment by rohitpaulk** [comment]
I found an easier way to validate this - I just hit the API directly and filtered by `doc_id`. There indeed is only one doc for `adam-test`

screen shot 2017-05-13 at 7 07 48 am

Looks like a CouchDB bug? 😃

**Comment by rohitpaulk** [comment]
http://docs.couchdb.org/en/2.0.0/api/database/changes.html doesn't say anything specific about this..

**Comment by rohitpaulk** [comment]
My JIRA skills have FINALLY come to use outside of companies I work for! https://issues.apache.org/jira/browse/COUCHDB-3418

**Comment by whit537** [comment]
Good sleuthing!

> Makes sense, I'm happy to go ahead with this

+1

**Comment by rohitpaulk** [comment]
Fake package inserted. 

```
→ heroku pg:psql -a gratipay
--> Connecting to postgresql-octagonal-73752
psql (9.6.2)
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

gratipay::DATABASE=> INSERT INTO packages (package_manager, name, description, emails) VALUES ('npm', 'adam-test', 'Dummy description', '{}');
INSERT 0 1
```

**Comment by rohitpaulk** [comment]
Hmm, it momentarily started to process a few docs - and then encountered an exception again. 

screen shot 2017-05-13 at 4 32 13 pm

Looking for the exception in Sentry...

**Comment by rohitpaulk** [comment]
The initial change we got stuck on was sequence `1908160`, now we're stuck on `1908415`

**Comment by rohitpaulk** [comment]
Hmm, I can't find the exception on Sentry! Tried using the exception reference number, that didn't work.

**Comment by rohitpaulk** [comment]
Exceptions fixed in #4468

**Comment by rohitpaulk** [comment]
This looks like yet another case of the same problem, this time the package name is `animatetypewriter`

**Comment by rohitpaulk** [comment]
I inserted a fake package for that, and we're stuck again! 

This time, at `protobufjs5-fixpackedfields` (seq: `1908498`)

**Comment by rohitpaulk** [comment]
I think we should go ahead with adding logs for deleted packages

File: tests/py/test_sync_npm.py (hunk 1 above)
Comment (whit537) [inline tests/py/test_sync_npm.py]: Gotcha. Before af41409c5bea0f3cea3afbd804e64c83ef97c693 this tested something other than our test function. Out of scope here but we could/should write a test or three for [`gratipay.cli.sync_npm.main`](https://github.com/gratipay/gratipay.com/blob/2080/gratipay/cli/sync_npm.py#L15-L40).

**Comment by whit537** [comment]
Alright, so our hypothesis is that this is either a bug or an unexpected feature of CouchDB, yes?

I guess in either case there's not much we can do but work around it, unsettling as it may be. Since both Papertrail and apparently now Sentry (#4468) are somewhat unreliable, does it make more sense to log an event in our own `events` table instead using `add_event`?

**Comment by rohitpaulk** [comment]
> Since both Papertrail and apparently now Sentry (#4468) are somewhat unreliable, does it make more sense to log an event in our own events table instead using add_event?

Yes, it does. Change incoming..

**Comment by rohitpaulk** [comment]
I _want_ to use [`add_event`](https://github.com/gratipay/gratipay.com/blob/5228f5c8bf2f4c8eaec34d2931c263a25df6d8b9/gratipay/application.py#L60), but it is part of `gratipay.Application`, and passing that down to `sync_npm` doesn't seem like the best idea, seems better when dependencies are explicit. I could instantiate `gratipay.Application` and then pass the `db` and `add_event` function separately, but passing functions between modules leads to confusion as far as I've observed. I see two viable options: 

- Move `add_event` to `utils`
- Create an `EventStorage` object, with an `add` function. Initiate it in `gratipay.Application`, and pass it down to `sync_npm`

**Comment by rohitpaulk** [comment]
Ah, wait.. I could just add it to the `Package` model as a class function.

**Comment by rohitpaulk** [comment]
Done.

**Comment by rohitpaulk** [comment]
Heard back from couchDB - this is intended. 

https://issues.apache.org/jira/browse/COUCHDB-3418

> Hi Paul,
> 
> This is expected. Database compression means that only the most recent revision required to successfully replicate the database is present. In the case of a deleted document, that means that only the "tombstone" (document ID + revision + "_deleted": true) is retained. This is done so that if the document is re-created later (with a revision 1 greater than the previous revision) we can successfully detect that this newer revision isn't, for instance, the original revision of that document. This is the same for all documents - older revisions will be cleaned away during database compression.

Also, from their docs: 

> Normal deletion of a document within CouchDB does not remove the document from the database, instead, the document is marked as _deleted=true (and a new revision is created). This is to ensure that deleted documents can be replicated to other databases as having been deleted. This also means that you can check the status of a document and identify that the document has been deleted by its absence.

That explains https://gratipay.slackarchive.io/gratipay/page-95/ts-1493996587423821 :)

**Comment by whit537** [comment]
Awesome. So do we still need to log deletions, if this is expected behavior?

**Comment by rohitpaulk** [comment]
> So do we still need to log deletions, if this is expected behavior?

Nope, we don't. Popping the [last commit](https://github.com/gratipay/gratipay.com/pull/4464/commits/08586b26414bf9414c843a5a74821b499f2f4d96) off...

**Comment by rohitpaulk** [comment]
Amended the comment text: 

screen shot 2017-05-17 at 9 04 10 am

**Pull request merged**